src/types/BeforeSwapDelta.sol

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {BalanceDelta} from "./BalanceDelta.sol";
+
+// Return type of the beforeSwap hook.
+// Upper 128 bits is the delta in specified tokens. Lower 128 bits is delta in unspecified tokens (to match the afterSwap hook)
+type BeforeSwapDelta is int256;
+
+// Creates a BeforeSwapDelta from specified and unspecified
+function toBeforeSwapDelta(int128 deltaSpecified, int128 deltaUnspecified)
+    pure
+    returns (BeforeSwapDelta beforeSwapDelta)
+{
+    /// @solidity memory-safe-assembly
+    assembly {
+        beforeSwapDelta := or(shl(128, deltaSpecified), and(0xffffffffffffffffffffffffffffffff, deltaUnspecified))
+    }
+}
+
+library BeforeSwapDeltaLibrary {
+    BeforeSwapDelta public constant ZERO_DELTA = BeforeSwapDelta.wrap(0);
+
+    /// extracts int128 from the upper 128 bits of the BeforeSwapDelta
+    /// returned by beforeSwap
+    function getSpecifiedDelta(BeforeSwapDelta delta) internal pure returns (int128 deltaSpecified) {
+        assembly {
+            deltaSpecified := shr(128, delta)
```

Comment:
`sar` ? Referencing this fix here I think for signed ints we want to shift right with sar https://github.com/Uniswap/v4-core/pull/586/files#diff-96131175180a7d11a6b3d18a431973bc72a649426f8ecf735f9e7ff973412c5a

Reply:
Maybe we need to fix this also in BalanceDelta ?

Reply:
good point. BalanceDelta was updated by @shuhuiluo in #641 so have used the same logic as that lib